### PR TITLE
util/deephash: handle slice edge-cases

### PR DESCRIPTION
**Pull request by dsnet**
It is unclear whether the lack of checking nil-ness of slices
was an oversight or a deliberate feature.
Lacking a comment, the assumption is that this was an oversight.

Also, expand the logic to perform cycle detection for recursive slices.
We do this on a per-element basis since a slice is semantically
equivalent to a list of pointers.

Signed-off-by: Joe Tsai <joetsai@digital-static.net>

### CODE REVIEW

File: util/deephash/deephash.go [hunk 1]
```diff
@@ -326,17 +331,34 @@ func makeSliceHasher(t reflect.Type) typeHasherFunc {
...
 	var hashElem typeHasherFunc
+	var isRecursive bool
 	init := func() {
 		hashElem = getTypeInfo(t.Elem()).hasher()
+		isRecursive = typeIsRecursive(t)
 	}
 
 	return func(h *hasher, p pointer) {
 		pa := p.sliceArray()
+		if pa.isNil() {
```
Comment (dsnet) [inline util/deephash/deephash.go:342]: If the lack of nil-checking is deliberate, I'll put a comment here to the effect.
The `deepprint` predecessor never did nil-checks for slices,
but it feels weird and inconsistent that we do nil checks for maps, but not slices.

At the end of the day, the semantics here is more governed by how other packages use this,
and I don't have the best insight into that.

**Comment by dsnet** [comment]
This is the last semantic change to `deephash` I have stacked up.
I might revamp the tests in the future, but I feel that `deephash` is in a good state after this.

**Review by bradfitz — APPROVED** [review]

File: util/deephash/deephash.go [hunk 2]
```diff
@@ -303,17 +308,38 @@ func makeSliceHasher(t reflect.Type) typeHasherFunc {
...
+					if idx, ok := h.visitStack.seen(pe.p); ok {
+						h.HashUint8(2) // indicates cycle
+						h.HashUint64(uint64(idx))
+						return
+					}
+					h.HashUint8(1) // indicates visiting slice element
+					h.visitStack.push(pe.p)
+					defer h.visitStack.pop(pe.p)
+					hashElem(h, pe)
+				}()
```
Comment (bradfitz) [inline util/deephash/deephash.go:340]: these immediately-called-funcs are kinda gross style.

just for the defer?

Comment (dsnet) [reply util/deephash/deephash.go:340]: Yea. I realized looking at the code earlier that it was buggy without it. Did you have an alternative suggestion?

Comment (bradfitz) [reply util/deephash/deephash.go:340]: Using method? Not using a defer?

Comment (dsnet) [reply util/deephash/deephash.go]: I realized that the recursive case is identical to normal hashing of elements, but with additional logic before.
Implemented it as a wrapper around the original `hashElem`.

**Comment by bradfitz** [comment]
LGTM

**Pull request merged**